### PR TITLE
Custom semaphore implementation

### PR DESCRIPTION
**Pull request by sporksmith**
This adds a reimplementation of libc's semaphore interface. Using it instead of libc's semaphore in our binary_spinning_semaphore results in the underlying `syscall` calls going through the shim's overridden `syscall` function, which is whitelisted in the seccomp filter when seccomp is in use. i.e. this allows us to avoid unnecessary seccomp traps in the shim <-> shadow IPC (without the previous workaround of whitelisting *all* futex operations).
 
We also use our implementation of libc's semaphore to override libc's semaphore in the shim.

Progress on #1454

### CODE REVIEW

File: src/lib/shim/shadow_sem.c
Comment (RWails) [reply src/lib/shim/shadow_sem.c:84]: Or, used in L146?

**Review by RWails — CHANGES_REQUESTED** [review]

File: src/lib/shim/shadow_sem.c
Comment (sporksmith) [reply src/lib/shim/shadow_sem.c:78]: > p.s. I'm noticing that you're not using <stdatomic.h> --- is this because we have to support GCC 4.8.5?

Right :/

> it might be worth a lightweight port to C++11 so that we can use the standard library.

Maybe, but then we'd definitely need an opaque wrapper type to expose to the C code, which you were hoping to get rid of :).

I do think using a more modern atomics api is probably worth it though - wdyt?

File: src/lib/shim/shadow_sem.c
Comment (sporksmith) [reply src/lib/shim/shadow_sem.c]: > Or, used in L146?

Right. Tried to clarify further in the type definition.

File: src/lib/shim/shadow_sem.c
Comment (sporksmith) [reply src/lib/shim/shadow_sem.c]: Good point. This shouldn't really happen, and there's no good way to recover. Added a comment and `panic` here instead of just returning an error.

File: src/lib/shim/shadow_sem.h
Comment (sporksmith) [reply src/lib/shim/shadow_sem.h:16]: Yeah, this was semi-cargo-culted from glibc. I agree that this level of encapsulation is probably overkill for our purposes. Leaving for now until we decide whether to convert to C++, in which case we *would* need this sort of type.

File: src/lib/shim/shadow_sem.c
Comment (RWails) [reply src/lib/shim/shadow_sem.c:78]: > I do think using a more modern atomics api is probably worth it though - wdyt?

Yep, you are right. The best compromise here is to use C++11's atomics in the standard library and declare an opaque struct for C. I think the change from `shadow_sem.c` to `shadow_sem.cc` shouldn't be too bad... agreed?

**Comment by RWails** [comment]
@sporksmith do you want me to take a stab at the `shadow_sem.c` to `.cc` rewrite? Or back to you?

**Comment by sporksmith** [comment]
> @sporksmith do you want me to take a stab at the `shadow_sem.c` to `.cc` rewrite? Or is the baton back in your hands?

@RWails I'm comfortable doing it, but if you have cycles and would like to take a stab at it I won't turn down the help :)

**Comment by robgjansen** [comment]
Wouldn't the effort required to write in cpp be better spent writing in rust?

**Comment by RWails** [comment]
> Wouldn't the effort required to write in cpp be better spent writing in rust?

@sporksmith WDYT? The C -> CPP translation would be almost completely substitutional (I think), whereas a Rust implementation would be more substantial. Maybe still worth it if the module will eventually be rewritten in Rust, anyways.

**Comment by sporksmith** [comment]
I agree it doesn't make sense to block this on a Rust rewrite right now. I probably should've written it in Rust in the first place, but I initially thought it'd only be a few lines of C :). (It was, until I saw there was a performance degradation and had to fix it by avoiding unnecessary FUTEX_WAKE operations).

Actually, I think we've settled on dropping support for Centos 7, in which case I can just use the C11 atomics here, which should be a quite simple change. I'll send out the "drop centos 7 support" PR out and once it's finalized I'll plan to circle back to this and convert either to C11 or C++ atomics assuming it's not much work.

If it ends up being more substantial then I'd suggest we move on with the current use of gcc intrinsic atomics  for the moment, and circle back to rewrite it in Rust.

**Comment by sporksmith** [comment]
Some mildly ad-hoc measurements for posterity, using shadow config:
```
general:
  stop_time: 4 s
  bootstrap_end_time: 4 s
network:
  graph:
    type: 1_gbit_switch
hosts:
  host:
    quantity: 10
    processes:
    - path: /home/jnewsome/projects/phantom-paper/experiments/mezzo-phold/phold/target/release/phold
      args: 10 host uniform 0 100 0 1337
      start_time: 1 s
```

@ main: using libc semaphore; getting good performance by allowing all `futex` syscalls in our seccomp filter; however that exception isn't compatible with running threaded virtual processes, since we need to emualte their `futex` syscalls.
```
$ hyperfine -M3 'cd `mktemp -d` && shadow /home/jnewsome/projects/phantom-paper/experiments/mezzo-phold/config.yaml --interpose-method=preload -p 12 > shadow.log'
Benchmark #1: cd `mktemp -d` && shadow /home/jnewsome/projects/phantom-paper/experiments/mezzo-phold/config.yaml --interpose-method=preload -p 12 > shadow.log
  Time (mean ± σ):     21.646 s ±  0.352 s    [User: 103.112 s, System: 72.352 s]
  Range (min … max):   21.355 s … 22.038 s    3 runs
```

@ main, with the (thread-incompatible) futex exception in the seccomp filter disabled; significant performance penalty due to going through our SIGSYS signal handler every time the shadow IPC semaphore makes a futex syscall:
```
$ hyperfine -M3 'cd `mktemp -d` && shadow /home/jnewsome/projects/phantom-paper/experiments/mezzo-phold/config.yaml --interpose-method=preload -p 12 > shadow.log'
Benchmark #1: cd `mktemp -d` && shadow /home/jnewsome/projects/phantom-paper/experiments/mezzo-phold/config.yaml --interpose-method=preload -p 12 > shadow.log
  Time (mean ± σ):     25.668 s ±  0.507 s    [User: 115.138 s, System: 93.088 s]
  Range (min … max):   25.140 s … 26.151 s    3 runs
```

this PR @ 73b13f2da6d9c7a63f0f955e771be446e1f26b4a ("use c11 atomics"): Switches the shadow IPC to use the semaphore implementation in the PR. Because it uses an interposable `syscall` *library* call, our seccomp filter knows to let it pass. This appears to get back to the original performance using the glibc semaphore, but without having to let all futex syscalls through in our seccomp filter.
```
$ hyperfine -M3 'cd `mktemp -d` && shadow /home/jnewsome/projects/phantom-paper/experiments/mezzo-phold/config.yaml --interpose-method=preload -p 12 > shadow.log'
Benchmark #1: cd `mktemp -d` && shadow /home/jnewsome/projects/phantom-paper/experiments/mezzo-phold/config.yaml --interpose-method=preload -p 12 > shadow.log
  Time (mean ± σ):     21.648 s ±  0.450 s    [User: 103.053 s, System: 72.515 s]
  Range (min … max):   21.145 s … 22.012 s    3 runs
```

this PR @ 73b13f2da6d9c7a63f0f955e771be446e1f26b4a ("use c11 atomics"), but without the 'nwaiters' opt. The first version I implemented looked more like this, but there is a slight performance penalty vs the glibc semaphore; probably enough to justify the extra complexity. It's a little tricky though, and I wouldn't be shocked if there's some incorrect detail that can lead to deadlock. In a pinch we can disable it by removing the `nwaiters` check in `shadow_sem_post`.
```
$ hyperfine -M3 'cd `mktemp -d` && shadow /home/jnewsome/projects/phantom-paper/experiments/mezzo-phold/config.yaml --interpose-method=preload -p 12 > shadow.log'
Benchmark #1: cd `mktemp -d` && shadow /home/jnewsome/projects/phantom-paper/experiments/mezzo-phold/config.yaml --interpose-method=preload -p 12 > shadow.log
  Time (mean ± σ):     22.592 s ±  0.504 s    [User: 107.018 s, System: 76.233 s]
  Range (min … max):   22.198 s … 23.160 s    3 runs
```

File: src/lib/shim/shadow_sem.h
Comment (sporksmith) [reply src/lib/shim/shadow_sem.h:27]: Done-ish. Unfortunately it turns out that C11 _Atomic is incompatible with C++, so I still had to hide that from the C++ code.

File: src/lib/shim/shadow_sem.c
Comment (sporksmith) [reply src/lib/shim/shadow_sem.c]: Done

File: src/lib/shim/shadow_sem.c
Comment (sporksmith) [reply src/lib/shim/shadow_sem.c]: Since we're dropping centos 7, I used C11 atomics, since I figured that'd be a little less work than changing to C++.

It turned out to be a little awkward with the design I had, because fields of `_Atomic` struct types can't be accessed without atomic operations. So e.g. after doing an atomic load of the whole semaphore, I still needed to do more atomic operations to access the futex sub-structure. Since those were already on a local copy I could use 'relaxed' operations, which I *think* are cheap (?), but it was still awkward.

I ended up rethinking it a bit and I think landed on a better design. I ended up using `memory_order_seq_cst` to ensure a consistent ordering between the operations on `_value` and `_nwaiters` instead of the previous approach of tracking the total number of wakes and jamming that into the futex. It's possible we could use a weaker memory order than seq_cst, but by my measurements we're already getting about the same performance as with the glibc semaphore, so I'm inclined to just leave it using that.

File: src/lib/shim/shadow_sem.h
Comment (RWails) [reply src/lib/shim/shadow_sem.h:27]: OK, so an alternative:

Both C and C++ define the `atomic_uint_fast32_t` typedef in <stdatomic.h> and <atomic>, respectively. Unfortunately, C doesn't define an ordinary `atomic_uint32_t` whereas C++ does. `atomic_uint_fast32_t` is guaranteed to be _at least_ 32 bytes, but could be more -_- I'm guessing in practice that `sizeof(atomic_uint_fast32_t) == 4` for every platform we care about.

So, we could use a single definition of the struct and use conditional compilation to grab the right header, and then use `static_assert()` to make sure that we're still getting the right size and alignment.

Tbh, I'm not sure this change is worth making, so I'll leave it up to you.

File: src/lib/shim/shadow_sem.c
Comment (RWails) [inline src/lib/shim/shadow_sem.c:18]: Are there checks you are missing in this predicate? Right now the function is a noop because the `static_assert()` calls are compiled away once the constant-expression condition is checked.

Comment (sporksmith) [reply src/lib/shim/shadow_sem.c]: Yes, this was intended as a noop and was just a function so that I could reference the assertions in both of the places they mattered, without duplicating them.

I changed this to be an accessor instead; hopefully that's clearer.

File: src/lib/shim/shadow_sem.c
Comment (RWails) [inline src/lib/shim/shadow_sem.c:87]: Shouldn't this be `int futex_rv` to match the return type of _futex_wake?

File: src/lib/shim/shadow_sem.h
Comment (sporksmith) [reply src/lib/shim/shadow_sem.h:27]: I couldn't get this to compile. It's a nice idea, but I think in the end it doesn't save us much since we still need the static asserts to validate we get the expected size and alignment.

File: src/lib/shim/shadow_sem.c
Comment (sporksmith) [reply src/lib/shim/shadow_sem.c]: Done

File: src/lib/shim/shadow_sem.c
Comment (RWails) [inline src/lib/shim/shadow_sem.c:16]: OK, last nits :)

1. I think we should drop the `const` on the parameter here and at Line 23 --- a `FUTEX_WAKE` operation will modify the futex word, right?

2. Based on current usage (i.e. immediately passing the returned address to `syscall`), I don't think this is necessary, but I think the return should `_Atomic` or at least `volatile`(?)

**Review by RWails — APPROVED** [review]
Looks great! A couple last comments but I think GTG 👍🏻

File: src/lib/shim/shadow_sem.c
Comment (sporksmith) [reply src/lib/shim/shadow_sem.c]: > I think we should drop the const on the parameter here and at Line 23 --- a FUTEX_WAKE operation will modify the futex word, right?

The FUTEX_WAKE doesn't modify the futex word; the waker modifies the word (with an atomic op) before doing the FUTEX_WAKE operation. Dropped the const anyway; see below.

> Based on current usage (i.e. immediately passing the returned address to syscall), I don't think this is necessary, but I think the return should _Atomic or at least volatile(?)

I changed this to keep the current pointer type instead of casting it. I ended up dropping the `const` for that to work (otherwise we need to cast away the `const` in the wake helper).

**Comment by sporksmith** [comment]
Thanks for the review!

**Pull request merged**
